### PR TITLE
[REF] Extract utility functions from ABI encoder.

### PR DESCRIPTION
**Pull request by chriseth**
This extracts all yul utility functions not directly related to abi encoding or decoding into a separate class such that they can also be used from other components. It also extracts the "only generate each function once" tool into its own class to be shared by the various function generating components.

### CODE REVIEW

File: libsolidity/codegen/MultiUseYulFunctionCollector.cpp [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+	return result;
+}
+
+string MultiUseYulFunctionCollector::createFunction(string const& _name, function<string ()> const& _creator)
+{
+	if (!m_requestedFunctions.count(_name))
+	{
+		string fun = _creator();
+		solAssert(!fun.empty(), "");
+		solAssert(fun.find("function " + _name) != string::npos, "Function not properly named.");
```
Comment (chriseth) [inline libsolidity/codegen/MultiUseYulFunctionCollector.cpp:48]: This assertion is the only addition in this PR :)

File: libsolidity/codegen/ABIFunctions.h [hunk 2]
```diff
@@ -116,27 +128,6 @@ class ABIFunctions
 		std::string toFunctionNameSuffix() const;
 	};
 
-	/// @returns the name of the cleanup function for the given type and
-	/// adds its implementation to the requested functions.
-	/// @param _revertOnFailure if true, causes revert on invalid data,
-	/// otherwise an assertion failure.
-	std::string cleanupFunction(Type const& _type, bool _revertOnFailure = false);
```
Comment (axic) [inline libsolidity/codegen/ABIFunctions.h]: Hm, can we ignore this part before #5815 is merged? I can work on it later the week, Friday earliest.

Comment (chriseth) [reply libsolidity/codegen/ABIFunctions.h]: Sure, I can just keep it in ABIFunctions for now.

**Comment by chriseth** [comment]
@axic updated

File: libsolidity/codegen/MultiUseYulFunctionCollector.cpp (hunk 1 above)
Comment (leonardoalt) [reply libsolidity/codegen/MultiUseYulFunctionCollector.cpp:48]: How does `fun` take `_name` into account?

Comment (chriseth) [reply libsolidity/codegen/MultiUseYulFunctionCollector.cpp:48]: It's a requirement imposed on the caller of `createFunction`.

Comment (leonardoalt) [reply libsolidity/codegen/MultiUseYulFunctionCollector.cpp:48]: Oh ok

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**